src/ui/carousel/carousel-synchronizer.ts

```diff
@@ -0,0 +1,187 @@
+import { Carousel } from './carousel';
+import { DefaultMap } from '../../map/default-map';
+
+/**
+ * Will synchronize the active slide index across multiple carousels.
+ *
+ * This can be useful if you have separate areas for assets, copy and/or
+ * navigation that need transition together.
+ *
+ * Example:
+ *
+ * DOM:
+ * ```
+ * <div class="carousel">
+ *   <div class="slide">1</div>
+ *   <div class="slide">2</div>
+ *   <div class="slide">3</div>
+ * </div>
+ * <div class="nav">
+ *   <div class="dot">1</div>
+ *   <div class="dot">2</div>
+ *   <div class="dot">3</div>
+ * </div>
+ * ```
+ *
+ * JS:
+ * ```
+ * const a = new Carousel(
+ *   document.querySelector('.carousel'),
+ *   Array.from(document.querySelectorAll('.container-a .slide')),
+ *   {
+ *     allowLooping: true,
+ *     transition: new DraggableSlide(),
+ *   });
+ * const b = new Carousel(
+ *   document.querySelector('.container-b'),
+ *   Array.from(document.querySelectorAll('.container-b .slide')),
+ *   {
+ *     allowLooping: true,
+ *   });
+ * const syncInstance = CarouselSynchronizer.getSingleton();
+ * syncInstance.sync(a, b);
+ *
+ * // At some point later, dispose.
+ * a.dispose();
+ * b.dispose();
+ * syncInstance.dispose(a);
+ * ```
+ *
+ * CSS:
+ * ```
+ * .carousel {
+ *   position: relative;
+ *   height: 20vh;
+ *   width: 80vw;
+ *   overflow: hidden;
+ *   display: flex;
+ *   flex-wrap: nowrap;
+ * }
+ * .slide {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ *   height: 100%;
+ *   width: 50%;
+ *   flex-shrink: 0;
+ *   font-size: 10vw;
+ *   user-select: none;
+ * }
+ * .slide:nth-child(1) {
+ *   background: paleturquoise;
+ * }
+ * .slide:nth-child(2) {
+ *   background: palevioletred;
+ * }
+ * .slide:nth-child(3) {
+ *   background: palegreen;
+ * }
+ * .nav {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ * }
+ * .dot {
+ *   flex-grow: 0;
+ *   flex-shrink: 0;
+ *   width: 10px;
+ *   height: 10px;
+ *   border-radius: 50%;
+ *   opacity: .25;
+ *   background: black;
+ *   transition: opacity 1s, transform 1s;
+ *   margin: 0 8px;
+ * }
+ * .dot.active {
+ *   transform: scale(1.25);
+ *   opacity: 1;
+ * }
+ * .dot.before, .dot.after {
+ *   transform: scale(0.9);
+ * }
+ * ```
+ */
+export class CarouselSynchronizer {
+  /**
+   * Returns a singleton.
+   */
+  static getSingleton(): CarouselSynchronizer {
```

Comment:
The change is doable, but the Carousel class is already quite large.

The more behaviour we encapsulate in a single class the harder it may become to debug the class in isolation. As it stands, if there is an issue with a carousel  syncing, we can remove the sync and test the carousel in isolation, narrowing down which file we need to be concerned with.

If the singleton pattern is the primary issue we can ditch it and do `new CarouselSynchronizer(...carousels)` instead, without giving up the debug/maintenance advantages of the separation of concerns.

2 or 3 would probably be the common case, Assets/Dots/Copy being synced. Some modules may need more, I have a current example that is synchronizing 7 carousels.

Reply:
Per side chat with Angus, I think the case to have to sync multiple levels of carousels is quite limited so I personally prefer remove the singleton usage and instead go with multiple instances of CarouselSynchronizer.

I think if the Carousel class is substantially large, we can just keep as a separate class.

Reply:
Kept CarouselSynchronizer separate, removed singleton requirement.